junifer/preprocess/fsl/apply_warper.py

```diff
@@ -0,0 +1,269 @@
+"""Provide class for warping via FSL FLIRT."""
+
+# Authors: Synchon Mandal <s.mandal@fz-juelich.de>
+# License: AGPL
+
+import subprocess
+import tempfile
+from pathlib import Path
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ClassVar,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Union,
+    cast,
+)
+
+import nibabel as nib
+import numpy as np
+
+from ...api.decorators import register_preprocessor
+from ...utils import logger, raise_error
+from ..base import BasePreprocessor
+
+
+if TYPE_CHECKING:
+    from nibabel import Nifti1Image
+
+
+@register_preprocessor
+class ApplyWarper(BasePreprocessor):
+    """Class for warping NIfTI images via FSL FLIRT.
+
+    Wraps FSL FLIRT ``applywarp``.
+
+    Parameters
+    ----------
+    ref : str
+        The data type to use as reference for warping.
+    on : str
+        The data type to use for warping.
+
+    Raises
+    ------
+    ValueError
+        If a list was passed for ``on``.
+
+    """
+
+    _EXT_DEPENDENCIES: ClassVar[
+        List[Dict[str, Union[str, bool, List[str]]]]
+    ] = [
+        {
+            "name": "fsl",
+            "optional": False,
+            "commands": ["applywarp"],
+        },
+    ]
+
+    def __init__(self, ref: str, on: str) -> None:
+        """Initialize the class."""
+        self.ref = ref
+        # Check only single data type is passed
+        if isinstance(on, list):
+            raise_error("Can only work on single data type, list was passed.")
+        self.on = on  # needed for the base validation to work
+        super().__init__(on=self.on)
+
+    def get_valid_inputs(self) -> List[str]:
+        """Get valid data types for input.
+
+        Returns
+        -------
+        list of str
+            The list of data types that can be used as input for this
+            preprocessor.
+
+        """
+        # Constructed dynamically
+        return [self.on, self.ref, "Warp"]
```

Comment:
This method should output the valid input that you can actually apply this transformer to. So in this case, is only BOLD (for the moment). Ref and Warp are requirements, but you can applywarp to the warp file.

Reply:
Check here: https://github.com/juaml/junifer/blob/e3e942324b2f90f103a8500105e123d664139737/junifer/preprocess/confounds/fmriprep_confound_remover.py#L213

Reply:
You need to override the `validate_input` method so it asks for all of the elements required.

Reply:
Do you think it's a good way to do it considering one can do `super().__init__(on="BOLD")`. Overriding it defeats the point of the base method imo, but again I don't want to be pedantic about it.

Reply:
This has been addressed as per our discussion to modify `BasePreprocessor`.